corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py

```diff
@@ -0,0 +1,734 @@
+        if user_history.message and user_history.change_messages:
+            raise Exception("got a migrated record")
+
+    # simply copy over changed_via to new column
+    # changed_via should always be present
+    user_history.changed_via = user_history.details.get('changed_via')
+    assert user_history.changed_via, f"Missing changed_via for {user_history.pk}"
+
+    # simply copy over user changes to new changes column
+    # changes might or might not be present
+    if user_history.details.get('changes'):
+        user_history.changes = user_history.details['changes']
+
+    if user_history.message:
+        change_messages = create_change_messages(user_history, skip_assertions)
+        assert change_messages, f"Change message not created for message {user_history.message}"
+        user_history.change_messages = change_messages
+
+    if save:
+        user_history.save()
+
+        migrated = migrate_user_history_to_log_entry(user_history)
+        assert migrated, f"Could not create log entry for User History record {user_history.pk}"
+
+
+def create_change_messages(user_history, skip_assertions):
+    """
+    A message is a ". " join of different messages.
+    So it should be possible to split by ". " and identify individual messages.
+    There are two messages that have ". " in it
+    - one for verified two factor disable
+        'Two factor disabled. Verified by: {verified_by}, verification mode: "{verification_mode}"'
+    - one for status update
+        'User {action}. Reason: "{reason}"'
+    They can be identified and handled accordingly
+    """
+
+    # split by . but not by something that starts with st like "St. " which was found in location names
+    # "St. " or "st. " was the only case found during the dry run
+    # if any message gets broken down poorly because of . in their messages,
+    # conversion won't work for such messages and they will be flagged
+    # all individual messages should have checks to ensure they are complete
+    messages = re.split(r"(?<!St)\. ", user_history.message, flags=re.IGNORECASE)
+    messages_not_converted = []
+    change_messages = {}
+
+    # flag for two factor disable message
+    disable_for_days = None
+
+    # flag for status update messages
+    status_update_active = None
+    status_update_reason = None
+
+    for message in messages:
+        if "Program: None" == message:
+            _check_for_double_entry(PROGRAM_FIELD, message, change_messages)
+
+            change_messages.update(
+                UserChangeMessage.program_change(None)
+            )
+
+        elif PROGRAM_REGEX.match(message):
+            _check_for_double_entry(PROGRAM_FIELD, message, change_messages)
+
+            program_name, program_id = PROGRAM_REGEX.match(message).groups()
+
+            assert program_name, f"Program name missing in message {message}"
+            assert program_id, f"Program id missing in message {message}"
+            program = Program.get(program_id)
```

Comment:
won't this raise a `ResourceNotFound` error if the doc isn't found?

Reply:
Yes, this would and all exceptions generated in `migrate` are caught and logged.

Do you mean the following assertion for program is hence not needed? That sounds right.